internal/delete/delete_item_action_handler_test.go

```diff
@@ -0,0 +1,264 @@
+			actions: map[*recordResourcesAction][]string{
+				new(recordResourcesAction).ForResource("po"): {"ns-1/pod-1", "ns-2/pod-2"},
+				new(recordResourcesAction).ForResource("pv"): {"pv-1", "pv-2"},
+			},
+		},
+		{
+			name:   "actions with selectors that don't match anything don't run for any resources",
+			backup: builder.ForBackup("velero", "velero").Result(),
+			tarball: test.NewTarWriter(t).
+				AddItems("pods", builder.ForPod("ns-1", "pod-1").Result()).
+				AddItems("persistentvolumeclaims", builder.ForPersistentVolumeClaim("ns-2", "pvc-2").Result()).
+				Done(),
+			apiResources: []*test.APIResource{test.Pods(), test.PVCs(), test.PVs()},
+			actions: map[*recordResourcesAction][]string{
+				new(recordResourcesAction).ForNamespace("ns-1").ForResource("persistentvolumeclaims"): nil,
+				new(recordResourcesAction).ForNamespace("ns-2").ForResource("pods"):                   nil,
+			},
+		},
+		{
+			name:   "single action with label selector runs only for those items",
+			backup: builder.ForBackup("velero", "velero").Result(),
+			tarball: test.NewTarWriter(t).
+				AddItems("pods", builder.ForPod("ns-1", "pod-1").ObjectMeta(builder.WithLabels("app", "app1")).Result(), builder.ForPod("ns-2", "pod-2").Result()).
+				AddItems("persistentvolumeclaims", builder.ForPersistentVolumeClaim("ns-1", "pvc-1").Result(), builder.ForPersistentVolumeClaim("ns-2", "pvc-2").ObjectMeta(builder.WithLabels("app", "app1")).Result()).
+				Done(),
+			apiResources: []*test.APIResource{test.Pods(), test.PVCs()},
+			actions: map[*recordResourcesAction][]string{
+				new(recordResourcesAction).ForLabelSelector("app=app1"): {"ns-1/pod-1", "ns-2/pvc-2"},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// test harness contains the fake API server/discovery client
+			h := newHarness(t)
+			for _, r := range tc.apiResources {
+				h.AddItems(t, r)
+			}
+
+			actions := []velero.DeleteItemAction{}
+			for action := range tc.actions {
+				actions = append(actions, action)
+			}
+
+			c := &Context{
+				Backup:          tc.backup,
+				BackupReader:    tc.tarball,
+				Filesystem:      fs,
+				DiscoveryHelper: h.discoveryHelper,
+				Actions:         actions,
+				Log:             log,
+			}
+
+			err := InvokeDeleteActions(c)
+			require.NoError(t, err)
+
+			for action, want := range tc.actions {
+				sort.Strings(want)
+				sort.Strings(action.ids)
+				assert.Equal(t, want, action.ids)
+			}
+		})
+	}
+}
+
+type harness struct {
+	*test.APIServer
+	discoveryHelper discovery.Helper
+}
+
+func newHarness(t *testing.T) *harness {
+	t.Helper()
+
+	apiServer := test.NewAPIServer(t)
+	log := logrus.StandardLogger()
+
+	discoveryHelper, err := discovery.NewHelper(apiServer.DiscoveryClient, log)
+	require.NoError(t, err)
+
+	return &harness{
+		APIServer:       apiServer,
+		discoveryHelper: discoveryHelper,
+	}
+}
+
+func (h *harness) AddItems(t *testing.T, resource *test.APIResource) {
```

Comment:
nit: Cleaner interface for this could be
```
func (h *harness) AddItems(t *testing.T, resource *[]test.APIResource)
```

Reply:
Also in https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore_test.go#L2821 this is not an exported method. Any reason to change that here?

Reply:
This is another opportunity to dedup code.
Suggest adding `TODOs` for us to do code dedup and cleanup as a follow-up.